protocol/jsm/v2/receiver.go

```diff
@@ -0,0 +1,166 @@
+/*
+ Copyright 2021 The CloudEvents Authors
+ SPDX-License-Identifier: Apache-2.0
+*/
+
+package jsm
+
+import (
+	"context"
+	"io"
+	"sync"
+
+	"github.com/nats-io/nats.go"
+
+	"github.com/cloudevents/sdk-go/v2/binding"
+	"github.com/cloudevents/sdk-go/v2/protocol"
+)
+
+type msgErr struct {
+	msg binding.Message
+}
+
+type Receiver struct {
+	incoming chan msgErr
+}
+
+func NewReceiver() *Receiver {
+	return &Receiver{
+		incoming: make(chan msgErr),
+	}
+}
+
+// MsgHandler implements nats.MsgHandler and publishes messages onto our internal incoming channel to be delivered
+// via r.Receive(ctx)
+func (r *Receiver) MsgHandler(msg *nats.Msg) {
+	r.incoming <- msgErr{msg: NewMessage(msg)}
+}
+
+func (r *Receiver) Receive(ctx context.Context) (binding.Message, error) {
+	select {
+	case msgErr, ok := <-r.incoming:
+		if !ok {
+			return nil, io.EOF
+		}
+		return msgErr.msg, nil
+	case <-ctx.Done():
+		return nil, io.EOF
+	}
+}
+
+type Consumer struct {
+	Receiver
+
+	Conn       *nats.Conn
+	Jsm        nats.JetStreamContext
+	Subject    string
+	Subscriber Subscriber
+	SubOpt     []nats.SubOpt
+
+	subMtx        sync.Mutex
+	internalClose chan struct{}
+	connOwned     bool
+}
+
+func NewConsumer(url, stream, subject string, natsOpts []nats.Option, jsmOpts []nats.JSOpt, subOpts []nats.SubOpt, opts ...ConsumerOption) (*Consumer, error) {
```

Comment:
when should I make a `NewConsumer` vs `NewReceiver`?

Reply:
NewConsumer is used by Nats subscriber，while  NewReceiver is used by NewConsumer， like `kafka_sarama` `nats``stan` and so on